ledger/participant-integration-api/src/main/scala/platform/apiserver/update/UpdatePathsTrie.scala

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.apiserver.update
+
+import collection.mutable
+import scala.annotation.tailrec
+
+object UpdatePathsTrie {
+  case class MatchResult(
+      isExact: Boolean,
+      matchedPath: UpdatePath,
+  ) {
+    def getUpdatePathModifier: UpdatePathModifier = matchedPath.modifier
+  }
+
+  def apply(
+      modifierO: Option[UpdatePathModifier],
+      nodes: (String, UpdatePathsTrie)*
+  ) = new UpdatePathsTrie(
+    nodes = mutable.SortedMap.from(nodes),
+    modifierO = modifierO,
+  )
+
+  def fromPaths(paths: Seq[UpdatePath]): Result[UpdatePathsTrie] = {
+    val t: Result[UpdatePathsTrie] = Right(UpdatePathsTrie(None))
+    paths.foldLeft(t)((trieResult, path) =>
+      for {
+        t <- trieResult
+        _ <- t.insertUniquePath(path)
+      } yield t
+    )
+  }
+}
+
+/** Data structure for storing and querying update paths.
+  *
+  * Each update path specifies:
+  * - a field path corresponding to a field of an update request proto message,
+  * - an update modifier.
+  *
+  * See also [[com.google.protobuf.field_mask.FieldMask]]).
+  *
+  * @param modifierO - non empty to signify a path ending in this node exists
+  */
+private[update] case class UpdatePathsTrie(
```

Comment:
I am not very fond of the fact that this is a mutable data-structure. Try to rephrase in a way that the accompanying object contains all the necessary recursive functions that generate an immutable instance

Reply:
I can do that, or use the mutable object to as a private builder for the immutable one.
Which approach do you prefer? (I'm slightly more inclined towards the mutable builder approach)

Reply:
fyi, implemented mutable builder approach